rfd/0038-database-access-aws-discovery.md

```diff
@@ -0,0 +1,163 @@
+when a new RDS, Aurora or Redshift database is added to the cluster, and attach
+it to the IAM user/role it is using.
+
+### Required permissions
+
+To support this, the database agent must have certain IAM permissions.
+
+- `iam.CreatePolicy` to be able to create policies for IAM RDS/Redshift
+  authentication.
+- `iam.AttachUserPolicy` or `iam.AttachRolePolicy` to be able to attach the
+  created policies to the IAM identity (user/role) it's running as.
+
+To know how to attach the policy to itself, the agent will use `sts.GetCallerIdentity`[1]
+call to determine IAM identity it's running as. This operation does not require
+any permissions.
+
+In addition, the agent will need to have `iam.DeletePolicy` and `iam.Detach*Policy`
+permissions to support deprovisioning of AWS databases removed from the cluster.
+
+### Privilege escalation
+
+Database agent being able to create and attach policies to itself raises
+privilege escalation concerns. Can somebody who has access to the database
+agent instance give itself unlimited IAM permissions?
+
+To prevent privilege escalation users can use IAM [permission
+boundaries](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html).
+Permission boundary is an IAM policy that does not grant any privileges by
+itself but defines the maximum privileges that can be granted to a user/role.
+For example, if permission boundary is set and doesn't explicitly allow to
+create EC2 instances, the user won't be able to create them even if their
+attached IAM policies allow it. The effective permissions set of an IAM identity
+is the intersection between their attached IAM policy and permission boundary.
+
+Applied to the database agent scenario, users will create a permission boundary
+that only allows the agent to create/attach IAM policies and connect to an RDS
+instance:
+
+```json
+{
+   "Version": "2012-10-17",
+   "Statement": [
+      {
+         "Effect": "Allow",
+         "Action": [
+           "iam:CreatePolicy",
+           "iam:DeletePolicy",
+           "iam:Attach*",
+           "iam:Detach*"
+         ],
+         "Resource": "*",
+      },
+      {
+         "Effect": "Allow",
+         "Action": [
+           "rds-db:connect"
+         ],
+         "Resource": [
+           "arn:aws:rds-db:*:1234567890:dbuser:*/*"
+         ]
+      }
+   ]
+}
+```
+
+With this permission boundary, whoever has access to the database agent instance
+will be able to create any policy but not use it for anything besides connecting
+to an RDS instance.
+
+## Discovery
+
+Database agent can auto-discover RDS and Aurora instances in the AWS account
+it has credentials for by matching resource tags. Auto-discovery is disabled
+by default. It can be enabled by providing a `selector` introduced in RFD 31 in
+the agent config:
+
+```yaml
+db_service:
+  enabled: "yes"
+  selectors:
+  - match_rds:
+      regions: ["us-east-1"]
```

Comment:
Should we provide an option to override the AWS account? Or do auto detect it based on where the DB node is running?

Reply:
Account is determined by the IAM credentials Teleport has. We could potentially extend this to support cross account access in future.